src/main/java/org/purescript/psi/expression/ExpressionIdentifierReference.kt

```diff
@@ -0,0 +1,79 @@
+import com.intellij.psi.util.parents
+import org.purescript.file.ExportedValuesIndex
+import org.purescript.psi.PSValueDeclaration
+
+class ExpressionIdentifierReference(expressionConstructor: PSExpressionIdentifier) :
+    LocalQuickFixProvider,
+    PsiReferenceBase.Poly<PSExpressionIdentifier>(
+        expressionConstructor,
+        expressionConstructor.qualifiedIdentifier.identifier.textRangeInParent,
+        false
+    ) {
+
+    override fun getVariants(): Array<Any> =
+        candidates.toList().toTypedArray()
+
+    override fun multiResolve(incompleteCode: Boolean): Array<ResolveResult> {
+        val name = element.name
+        val declaration = candidates.firstOrNull { it.name == name }
+            ?: return emptyArray()
+        // TODO We may not want to have a PsiPolyVariantReference reference here
+        return if (declaration is PSValueDeclaration) {
+            val module = declaration.module
+            val allValueDeclarations = candidates.filterIsInstance<PSValueDeclaration>()
+                .filter { it.name == name }
+                .filter { it.module == module }
+                .toMutableList()
+            createResults(allValueDeclarations)
+        } else {
+            createResults(declaration)
+        }
+    }
+
+    private val candidates: Sequence<PsiNamedElement>
+        get() {
+            val module = element.module ?: return emptySequence()
+            val qualifyingName = element.qualifiedIdentifier.moduleName?.name
+            if (qualifyingName != null) {
+                val importDeclaration = module.importDeclarations
+                    .firstOrNull { it.importAlias?.name == qualifyingName }
+                    ?: return emptySequence()
+                return sequence {
+                    yieldAll(importDeclaration.importedValueDeclarations)
+                    yieldAll(importDeclaration.importedForeignValueDeclarations)
+                    // TODO Support imported class members
+                }
+            } else {
+                return sequence {
+                    for (parent in element.parents) {
+                        when (parent) {
+                            is PSValueDeclaration -> yieldAll(parent.varBindersInParameters.values)
+                        }
+                    }
+                    // TODO Support values defined in the expression
+                    yieldAll(module.valueDeclarations.toList())
+                    yieldAll(module.foreignValueDeclarations.toList())
+                    // TODO Support local class members
+                    val importDeclarations = module.importDeclarations.filter { it.importAlias == null }
+                    yieldAll(importDeclarations.flatMap { it.importedValueDeclarations })
+                    yieldAll(importDeclarations.flatMap { it.importedForeignValueDeclarations })
+                    // TODO Support imported class members
+                }
+            }
+        }
```

Comment:
This code is a little messy, but we need to handle the case of qualified vs. non-qualified differently. The `for..when` may become bigger when we include support for e.g. `let` and `where` statements.  The todos are so that it's clear that we don't support class member navigation yet.